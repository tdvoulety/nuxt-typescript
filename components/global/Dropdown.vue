<template>
  <div class="dropdown" :class="{ 'show': opened, 'disabled': disabled }">
    <button type="button" class="dropdown__toggle" :id="dropdownId" :disabled="disabled">
      <span :id="dropdownTextId">{{ text }}</span>
    </button>

    <ul role="menu" tabindex="-1" class="dropdown__menu" :class="{ 'show': opened }" @click.stop>
      <li v-for="item in items" :key="item.value">
        <button type="button" class="dropdown__item" @click="handleClickOnItem(item)">
          {{ item.text }}
        </button>
      </li>
    </ul>
  </div>
</template>

<script>
import { GLOBAL_CLICK_EVENT } from '@/constants/global'

export default {
  props: {
    text: {
      type: [String, Number],
      required: true,
      default: '',
    },
    items: {
      type: Array,
      required: true,
      default: () => [],
    },
    disabled: {
      type: Boolean,
      required: false,
      default: false,
    },
  },
  data() {
    return {
      opened: false,
      id: null,
    }
  },
  computed: {
    dropdownId() {
      return `dropdown-${this.id}`
    },
    dropdownTextId() {
      return `${this.dropdownId}__text`
    },
  },
  methods: {
    toggleMenu() {
      this.opened = !this.opened
    },
    handleGlobalClickEvent(e) {
      if (!this.disabled && e && (e.target.id === this.dropdownId || e.target.id === this.dropdownTextId)) {
        this.toggleMenu()
        e.stopPropagation()
      } else {
        this.hideMenu()
      }
    },
    hideMenu() {
      this.opened = false
    },
    handleClickOnItem(item) {
      this.hideMenu()
      this.$emit('onClick', item)
    },
  },
  mounted() {
    // eslint-disable-next-line no-underscore-dangle
    this.id = this._uid
    this.$root.$on(GLOBAL_CLICK_EVENT, this.handleGlobalClickEvent)
  },
  destroyed() {
    this.$root.$off(GLOBAL_CLICK_EVENT, this.handleGlobalClickEvent)
  },
}
</script>

<style scoped lang="scss">
.dropdown {
  position: relative;
  vertical-align: middle;

  &__toggle {
    font-family: 'Source Sans Pro', sans-serif;
    font-size: 1.4rem;
    white-space: nowrap;
    outline: none;
    cursor: pointer;
    margin: 0;
    padding: 0;
    border: 0;
    background: 0;
    color: inherit;

    &:hover, &:focus {
      outline: none;
    }

    &:after {
      display: inline-block;
      margin-left: .255em;
      vertical-align: .1em;
      content: '';
      border-top: .3em solid;
      border-right: .3em solid transparent;
      border-bottom: 0;
      border-left: .3em solid transparent;
    }
  }

  &__menu {
    position: absolute;
    top: 100%;
    left: auto;
    right: 0;
    bottom: auto;
    z-index: 1000;
    display: none;
    float: left;
    min-width: 10rem;
    padding: .5rem 0;
    margin: .125rem 0 0;
    font-size: 1rem;
    text-align: left;
    list-style: none;
    background-color: #fff;
    background-clip: padding-box;
    border: 1px solid rgba(0,0,0,.15);
    border-radius: .4rem;

    &.show {
      display: block;
    }
  }

  &__item {
    font-family: 'Source Sans Pro', sans-serif;
    display: block;
    width: 100%;
    padding: .25rem 1.5rem;
    clear: both;
    font-size: 1.4rem;
    text-align: inherit;
    white-space: nowrap;
    background-color: transparent;
    border: 0;
    cursor: pointer;

    &:hover {
      background: grey;
      color: currentColor;
    }

    &:focus, &:hover {
      text-decoration: none;
      background-color: #ecedee;
      outline: none;
    }
  }

  &.disabled {

    .dropdown__toggle {
      cursor: default;

      &:after {
        opacity: .5;
      }
    }
  }
}
</style>
