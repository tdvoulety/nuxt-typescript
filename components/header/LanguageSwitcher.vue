<template>
  <div class="language-switcher">
    <div v-if="selectedLocale" class="language-switcher__current">
      <span>{{ selectedLocale.name }}</span>
    </div>

    <Dropdown
      :text="(selectedLocale && selectedLocale.name) || ''"
      :items="links"
      @onClick="handleClickOnLocale"
    />
  </div>
</template>

<script>
export default {
  components: {
    Dropdown: () => import('@/components/global/Dropdown'),
  },
  computed: {
    locales() {
      const { config: { i18n: { locales } } } = this.$store.state

      return locales
    },
    selectedLocale() {
      if (this.locales) {
        return this.locales.find((item) => item.code === this.$i18n.locale)
      }

      return null
    },
    links() {
      if (this.locales && this.selectedLocale) {
        const filteredLocales = this.locales.filter((item) => item.code !== this.selectedLocale.code)

        return filteredLocales.map((item) => {
          const isDev = process.env.NODE_ENV !== 'production'
          const port = isDev ? `:${process.env.DEV_PORT || 3000}` : ''

          return {
            code: item.code,
            value: `http://${item.domain}${port}`,
            text: item.name,
          }
        })
      }

      return []
    },
  },
  methods: {
    handleClickOnLocale({ value }) {
      window.location.assign(value)
    },
  },
}
</script>

<style lang="scss" scoped>
.language-switcher {

  ::v-deep .dropdown {

    &__toggle, &__toggle:active, &__toggle:focus {
      color: grey;
      background: transparent;
      border: 0;
      font-size: 1.4rem;
      line-height: 2.8rem;
      padding: 0 1rem;

      &:after {
        margin-left: .4em;
        vertical-align: .155em;
      }
    }

    &__item {
      font-size: 1.4rem;
      line-height: 1.428571429;
      padding: 0;

      @include media-breakpoint-up(md) {
        padding: .25rem 2.5rem;
        line-height: 1.428571429;

        &:hover {
          text-decoration: none;
        }
      }
    }
  }

  &__current {
    display: block;
    margin: .5rem 1.5rem;
    span {
      display: block;
      padding: 1.5rem;
      font-size: 1.8rem;
      text-align: left;
      background: grey;
      color: grey;
      border-radius: .4rem;
    }
    @include media-breakpoint-up(md) {
      display: none;
    }
  }
}
</style>
