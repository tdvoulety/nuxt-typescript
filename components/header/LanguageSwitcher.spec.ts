// @ts-nocheck
import LanguageSwitcher from '@/components/header/LanguageSwitcher'
import Dropdown from '@/components/global/Dropdown'
import { mountComponent, rendersToSnapshot } from '~/test/unit/helpers'

const mount = ({ mocks = {} } = {}) => mountComponent(LanguageSwitcher, {
  state: {
    config: {
      i18n: {
        defaultLocale: 'cs',
        locales: [
          {
            code: 'cs',
            name: 'Česky',
            domain: 'placestoreshop.cz',
          },
          {
            code: 'sk',
            name: 'Slovensky',
            domain: 'placestoreshop.sk',
          },
        ],
      },
    },
  },
  stubs: {
    Dropdown,
  },
  mocks,
})

describe('LanguageSwitcher', () => {
  it('render selected locale if locale is selected', () => {
    const locale = {
      code: 'cs',
      name: 'Česky',
      domain: 'placestoreshop.cz',
    }
    const { wrapper } = mount({ mocks: { $i18n: { locale: locale.code } } })

    expect(wrapper.findAll('.language-switcher').length).toEqual(1)
    expect(wrapper.findAll('.language-switcher__current').length).toEqual(1)
    expect(wrapper.vm.selectedLocale).toEqual(locale)
    expect(wrapper.find('.language-switcher__current span').text()).toBe(locale.name)
  })

  it('don\'t render selected locale if locale is not selected', () => {
    const { wrapper } = mount({ mocks: { $i18n: { locale: null } } })

    expect(wrapper.findAll('.language-switcher').length).toEqual(1)
    expect(wrapper.findAll('.language-switcher__current').length).toEqual(0)
  })

  it('renders to a snapshot', () => {
    const { wrapper } = mount()

    rendersToSnapshot(wrapper)
  })
})
