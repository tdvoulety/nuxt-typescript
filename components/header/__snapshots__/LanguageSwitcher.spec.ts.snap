// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSwitcher renders to a snapshot 1`] = `
<div class="language-switcher">
  <div class="language-switcher__current"><span>Česky</span></div>
  <div class="dropdown"><button type="button" id="dropdown-null" class="dropdown__toggle"><span id="dropdown-null__text">Česky</span></button>
    <ul role="menu" tabindex="-1" class="dropdown__menu">
      <li><button type="button" class="dropdown__item">
          Slovensky
        </button></li>
    </ul>
  </div>
</div>
`;
